File: rate-limiter/src/main/java/com/ratelimiter/algorithm/TokenBucket.java
package com.ratelimiter.algorithm;

import com.ratelimiter.domain.RateLimitKey;
import com.ratelimiter.domain.RateLimitResult;
import com.ratelimiter.domain.RateLimitRule;
import com.ratelimiter.repository.RateLimitRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Component;

import java.time.Instant;

/**
 * Token Bucket rate limiting algorithm implementation.
 *
 * <p>Tokens are added to a bucket at a fixed rate. Each request consumes one token.
 * Requests are rejected when the bucket is empty. The bucket has a maximum capacity
 * which allows for controlled bursting.
 *
 * <p>Configuration:
 * <ul>
 *   <li>maxRequests: Maximum bucket capacity (burst size)</li>
 *   <li>windowSize: Time to fully refill the bucket</li>
 *   <li>Refill rate: maxRequests / windowSize tokens per second</li>
 * </ul>
 */
@Slf4j
@Component
@RequiredArgsConstructor
public class TokenBucket implements RateLimitAlgorithm {

    private final RateLimitRepository repository;

    @Override
    public AlgorithmType getType() {
        return AlgorithmType.TOKEN_BUCKET;
    }

    @Override
    public RateLimitResult checkAndIncrement(RateLimitKey key, RateLimitRule rule, int weight) {
        long now = System.currentTimeMillis();

        // Get current bucket state
        TokenBucketState state = repository.getTokenBucketState(key, rule);

        // Calculate tokens to add based on elapsed time
        long elapsedMs = now - state.lastRefillTime();
        double refillRate = (double) rule.maxRequests() / (rule.windowSizeSeconds() * 1000);
        long tokensToAdd = (long) (elapsedMs * refillRate);

        // Refill bucket (capped at max capacity)
        long availableTokens = Math.min(rule.maxRequests(), state.tokens() + tokensToAdd);

        // Check if enough tokens available
        if (availableTokens < weight) {
            log.debug("Rate limit exceeded for key {}: {} tokens available, {} required",
                key, availableTokens, weight);

            // Calculate when enough tokens will be available
            long tokensNeeded = weight - availableTokens;
            long msUntilAvailable = (long) (tokensNeeded / refillRate);
            Instant resetTime = Instant.ofEpochMilli(now + msUntilAvailable);

            return RateLimitResult.rejected(
                rule.maxRequests() - availableTokens, // Current usage
                rule.maxRequests(),
                resetTime,
                rule.id(),
                rule.name()
            );
        }

        // Consume tokens
        long newTokens = availableTokens - weight;
        repository.setTokenBucketState(key, rule, new TokenBucketState(newTokens, now));

        // Reset time is when bucket would be full again
        long msUntilFull = (long) ((rule.maxRequests() - newTokens) / refillRate);
        Instant resetTime = Instant.ofEpochMilli(now + msUntilFull);

        return RateLimitResult.allowed(
            rule.maxRequests() - newTokens, // Current usage
            rule.maxRequests(),
            resetTime
        );
    }

    @Override
    public RateLimitResult check(RateLimitKey key, RateLimitRule rule) {
        long now = System.currentTimeMillis();

        TokenBucketState state = repository.getTokenBucketState(key, rule);

        long elapsedMs = now - state.lastRefillTime();
        double refillRate = (double) rule.maxRequests() / (rule.windowSizeSeconds() * 1000);
        long tokensToAdd = (long) (elapsedMs * refillRate);
        long availableTokens = Math.min(rule.maxRequests(), state.tokens() + tokensToAdd);

        long msUntilFull = (long) ((rule.maxRequests() - availableTokens) / refillRate);
        Instant resetTime = Instant.ofEpochMilli(now + msUntilFull);

        if (availableTokens < 1) {
            return RateLimitResult.rejected(
                rule.maxRequests() - availableTokens,
                rule.maxRequests(),
                resetTime,
                rule.id(),
                rule.name()
            );
        }

        return RateLimitResult.allowed(
            rule.maxRequests() - availableTokens,
            rule.maxRequests(),
            resetTime
        );
    }

    @Override
    public long getCurrentCount(RateLimitKey key, RateLimitRule rule) {
        long now = System.currentTimeMillis();
        TokenBucketState state = repository.getTokenBucketState(key, rule);

        long elapsedMs = now - state.lastRefillTime();
        double refillRate = (double) rule.maxRequests() / (rule.windowSizeSeconds() * 1000);
        long tokensToAdd = (long) (elapsedMs * refillRate);
        long availableTokens = Math.min(rule.maxRequests(), state.tokens() + tokensToAdd);

        return rule.maxRequests() - availableTokens;
    }

    @Override
    public void reset(RateLimitKey key, RateLimitRule rule) {
        repository.setTokenBucketState(key, rule, new TokenBucketState(rule.maxRequests(), System.currentTimeMillis()));
    }

    /**
     * Represents the state of a token bucket.
     */
    public record TokenBucketState(long tokens, long lastRefillTime) {
        public static TokenBucketState initial(int maxTokens) {
            return new TokenBucketState(maxTokens, System.currentTimeMillis());
        }
    }
}

File: rate-limiter/src/main/resources/scripts/token_bucket.lua
-- Token Bucket Rate Limiting Script
--
-- KEYS[1] = bucket key (stores tokens:last_refill_time)
-- ARGV[1] = max tokens (bucket capacity)
-- ARGV[2] = refill rate (tokens per second)
-- ARGV[3] = current timestamp (milliseconds)
-- ARGV[4] = tokens to consume (default 1)
-- ARGV[5] = TTL in seconds
--
-- Returns: {allowed (0/1), tokens_remaining, tokens_consumed, reset_time_ms}

local bucket_key = KEYS[1]
local max_tokens = tonumber(ARGV[1])
local refill_rate = tonumber(ARGV[2])
local now = tonumber(ARGV[3])
local consume = tonumber(ARGV[4]) or 1
local ttl = tonumber(ARGV[5])

-- Get current bucket state
local bucket = redis.call('HMGET', bucket_key, 'tokens', 'last_refill')
local tokens = tonumber(bucket[1])
local last_refill = tonumber(bucket[2])

-- Initialize bucket if it doesn't exist
if tokens == nil then
    tokens = max_tokens
    last_refill = now
end

-- Calculate tokens to add based on elapsed time
local elapsed_ms = now - last_refill
local tokens_to_add = elapsed_ms * refill_rate / 1000

-- Refill bucket (capped at max capacity)
tokens = math.min(max_tokens, tokens + tokens_to_add)

-- Check if enough tokens available
if tokens < consume then
    -- Rejected - not enough tokens
    -- Calculate when enough tokens will be available
    local tokens_needed = consume - tokens
    local ms_until_available = tokens_needed * 1000 / refill_rate
    local reset_time = now + ms_until_available

    -- Update last_refill time even on rejection
    redis.call('HSET', bucket_key, 'tokens', tokens, 'last_refill', now)
    redis.call('EXPIRE', bucket_key, ttl)

    return {0, math.floor(tokens), consume, math.floor(reset_time)}
end

-- Consume tokens
local new_tokens = tokens - consume

-- Save new bucket state
redis.call('HSET', bucket_key, 'tokens', new_tokens, 'last_refill', now)
redis.call('EXPIRE', bucket_key, ttl)

-- Calculate reset time (when bucket would be full)
local ms_until_full = (max_tokens - new_tokens) * 1000 / refill_rate
local reset_time = now + ms_until_full

return {1, math.floor(new_tokens), consume, math.floor(reset_time)}
